pkg/operation/common/utils.go

```diff
@@ -166,6 +166,24 @@ func ProjectForNamespace(projectLister gardencorelisters.ProjectLister, namespac
 	return nil, apierrors.NewNotFound(gardencorev1beta1.Resource("Project"), fmt.Sprintf("for namespace %s", namespaceName))
 }
 
+// ProjectForNamespaceWithClient returns the project object responsible for a given <namespace>.
+// It tries to identify the project object by looking for the namespace name in the project spec.
+func ProjectForNamespaceWithClient(ctx context.Context, c client.Client, namespaceName string) (*gardencorev1beta1.Project, error) {
+	projectList := &gardencorev1beta1.ProjectList{}
+	err := c.List(ctx, projectList)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, project := range projectList.Items {
+		if project.Spec.Namespace != nil && *project.Spec.Namespace == namespaceName {
+			return &project, nil
+		}
+	}
+
+	return nil, apierrors.NewNotFound(gardencorev1beta1.Resource("Project"), fmt.Sprintf("for namespace %s", namespaceName))
```

Comment:
Does it make sense to put this into a separate function that can be called from both `ProjectForNamespace` and `ProjectForNamespaceWithClient`?

Reply:
Yes, it would make sense. But I don't see an easy and clean solution how :/
As `projectLister.List` returns `[]*gardencorev1beta1.Project` and `client.List()` writes to a `*gardencorev1beta1.ProjectList{}` which contains a `[]gardencorev1beta1.Project`, this helper function would need to accept a `interface{}` param, which would have to be casted again. Other option would be to copy all items of a list in one function to a slice of a type accepted by the helper function...

Reply:
Yes, I actually meant following the second approach you mentioned. I think I have done something similar in the `ControllerRegistration` controller. It's not much code, but this way the "business logic" would be captured in a single function and you only have to transform the input (project lists) to satisfy this function's signature.

Reply:
Well, that makes sense. Done :)